src/TorchSharp/AutogradFunction.cs

```diff
@@ -0,0 +1,502 @@
+                private bool _materializeGrads = true;
+                private List<Tensor> _tensorsToSave = new();
+                private List<Tensor> _dirtyTensors = new();
+                private List<Tensor> _nonDifferentiableTensors = new();
+                private List<SavedVariable> _savedVariables = new();
+                private Dictionary<string, object> _savedData = new();
+                
+                internal AutogradContext(Node node)
+                {
+                    _node = node;
+                }
+
+                /// <summary>
+                /// Saves the list of variables for a future call to `backward`. This
+                /// should be called at most once from inside of `forward`.
+                /// </summary>
+                public void save_for_backward(List<Tensor> tensors) => _tensorsToSave = tensors;
+
+                /// <summary>
+                /// Saves a non tensor object in the data field for a future call to `backward`.
+                /// </summary>
+                public void save_data(string key, object value)
+                {
+                    if (value is Tensor)
+                        Console.WriteLine("Attempted to save a tensor object in the data object. Please store tensors by calling `save_for_backward()`");
+                    _savedData[key] = value;
+                }
+
+                /// <summary>
+                /// Marks variables in the list as modified in an in-place operation. This
+                /// should be called at most once from inside of `forward` and all arguments
+                /// should be inputs.
+                /// </summary>
+                public void mark_dirty(List<Tensor> tensors) => _dirtyTensors = tensors.ToList();
+
+                /// <summary>
+                /// Marks outputs in the list as not requiring gradients. This should be
+                /// called at most once from inside of `forward` and all arguments should be
+                /// outputs.
+                /// </summary>
+                public void mark_non_differentiable(List<Tensor> tensors) => _nonDifferentiableTensors = tensors.ToList();
+
+                /// <summary>
+                /// Sets whether undefined output grad tensors should be expanded to tensors
+                /// full of zeros before calling backward function. Default value is true.
+                /// </summary>
+                public void set_materialize_grads(bool value) => _materializeGrads = value;
+
+                public bool MaterializeGrads => _materializeGrads;
+
+                /// <summary>
+                /// Get the list of variables that were saved in `forward` using
+                /// `save_for_backward()`. Before returning them to the user, a check is made
+                /// to ensure that they were not modified by any in-place operations.
+                /// </summary>
+                public List<Tensor> get_saved_variables()
```

Comment:
When I call 'zero_grad()' on the optimizer I'm using for training, can the saved variables be cleared? Are they?

Reply:
Yup! When you call zero_grad then the node is disposed, which then in turn disposes the saved variables.